---
title: "Epignetic Cancer project"
output:                                                                    
  html_document:                                                           
    toc: true                                                              
    number_sections: true                                                  
    code_folding: "hide"                                                   
    toc_float:                                                             
      collapse: false                                                      
      smooth_scroll: false
    css: styles.css 
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, out.width = "1250px", out.height = "625px")
```


# RNA-Seq
## Quality checks
### Alignment statistics
```{r}
knitr::include_graphics("../pdf/RNA_alignment_statistics.pdf")
```
### PCC replicates
```{r}
knitr::include_graphics("../pdf/RNA_PCC_replicates.pdf")
```
### PCA replicates (using norm counts)
```{r}
knitr::include_graphics("../pdf/RNA_PCA_counts.pdf")
```
### GFP reads (genetic system validation)
```{r}
knitr::include_graphics("../pdf/RNA_GFP_norm_counts.pdf")
```
## Compare GFP and noGFP systems
### PCA between conditions (using log2FC)
```{r}
knitr::include_graphics("../pdf/RNA_PCA_log2FC.pdf")
```
### Overlaps between up and down genes 
```{r}
knitr::include_graphics("../pdf/RNA_chisq_overlap_GFP_noGFP_systems.pdf")
```
## MA plots
RNA-Seq differentially expressed genes

```{r}
knitr::include_graphics("../pdf/RNA_MA_plots_RNA.pdf")
```

## FC up vs down genes per cdition
```{r}
knitr::include_graphics("../pdf/RNA_boxplot_FC_up_down_genes.pdf")
```

## Alluvial plots
Visualize transitions between conditions

```{r}
knitr::include_graphics("../pdf/RNA_alluvial_plot_timecourse_RNA.pdf")
```
## Gene overlaps between conditions
```{r}
knitr::include_graphics("../pdf/RNA_upsetplot_overlapping_genes.pdf")
```
## Heatmap logFC genes of interest
```{r}
knitr::include_graphics("../pdf/RNA_heatmap_FC_GOF.pdf")
```
## GO up and down genes per condition
```{r}
knitr::include_graphics("../pdf/RNA_GO_up_down_genes_GFP-.pdf")
```




# RNA-Seq clusters analyses
## Cluster figure (heatmap, STRING network, GO, motifs)
```{r}
knitr::include_graphics("../pdf/cluster_GFP-_system_RNA.pdf")
```
## STRING newtork per cluster
```{r}
knitr::include_graphics("../pdf/cluster_network_per_cluster.pdf")
```
## Boxplot quantif FPKM and enrichments 
Stratified per cluster and PRC1 binding

```{r}
knitr::include_graphics("../pdf/cluster_PRC1_bound_unbound_RNA.pdf")
```
## Check percentage of PRC1 bound genes
```{r}
knitr::include_graphics("../pdf/cluster_percentage_PRC1_bound_genes.pdf")
```

## Tissue specificity score 
Stratified per cluster and PRC1 binding

```{r}
knitr::include_graphics("../pdf/cluster_PRC1_bound_unbound_tissue_specifity.pdf")
```




# CUTNRUN
## Quality Checks
### Alignment statistics
```{r}
knitr::include_graphics("../pdf/cutnrun_alignment_statistics_cutnrun.pdf")
```
### PCC between replicates
```{r}
knitr::include_graphics("../pdf/cutnrun_intersect_diff_regions.pdf")
```
### PCC between replicates
```{r}
knitr::include_graphics("../pdf/cutnrun_PCC_replicates.pdf")
```
## MA plots (DESeq2)
```{r, , out.height = "1800px"} 
knitr::include_graphics("../pdf/cutnrun_MA_plots.png")
```




# Reverting vs non reverting genes
## Defining characteristics
### FC between conditions
```{r}
knitr::include_graphics("../pdf/recovery_FC_conditions.pdf")
```
### Overlap with RNA clusters
```{r}
knitr::include_graphics("../pdf/recovery_overlap_RNA_clusters.pdf")
```
### Chromhmm classes
```{r}
knitr::include_graphics("../pdf/recovery_chromhmm_class.pdf")
```
### Overlap with K27 and K118 domains
```{r}
knitr::include_graphics("../pdf/recovery_overlap_K27_K118_domains.pdf")
```
### Size of overlapping K27 domains
```{r}
knitr::include_graphics("../pdf/recovery_size_K27_domains.pdf")
```
### FPKMs between cditions and percentile 
```{r}
knitr::include_graphics("../pdf/recovery_groups_FPKMs.pdf")
```
### GO
```{r}
knitr::include_graphics("../pdf/recovery_GOs.pdf")
```
### STRING network
```{r}
knitr::include_graphics("../pdf/recovery_STRING_network.pdf")
```
### Motifs enrichment
#### Over represented motifs
```{r}
knitr::include_graphics("../pdf/recovery_motifs_enrichment.pdf")
```
#### Heatmap ZEB and STAT motifs
```{r}
knitr::include_graphics("../pdf/recovery_promoters_motif_counts_heatmap.pdf")
```

## Average tracks 
### Gene promoters PH18/WT conditions
```{r}
knitr::include_graphics("../pdf/recovery_ph18_average_tracks.pdf")
```
### Gene promoters between RNAi condition
```{r}
knitr::include_graphics("../pdf/recovery_cditions_average_tracks.pdf")
```
### PH assigned peaks 
PH peaks assigned to their closest gene -> split between recovery & noRecovery

Not used anymore

### Compare PH peaks from Science Advances
PH peaks from the Science Advances were used -> aTSS, enhancer, peaks overlapping recovering/non recovering genes promoters

```{r}
knitr::include_graphics("../pdf/recovery_neoPRC1_average_tracks.pdf")
```

## Cutnrun FC between KD and PH18
```{r}
knitr::include_graphics("../pdf/recovery_ChIP_FC_cditions.pdf")
```
## K36me3 at noRecovery vs activity matched genes
```{r}
knitr::include_graphics("../pdf/recovery_H3K36me3_noRecovery_genes.pdf")
```
## K27Ac at recovery/noRecovery vs activity matched genes
```{r}
knitr::include_graphics("../pdf/recovery_H3K27Ac_activity_matched_genes.pdf")
```
## Compare FC at Day11 with PH  K27  K118 enrichments
```{r}
knitr::include_graphics("../pdf/recovery_PH_levels_vs_FC_D9_D11.pdf")
```


# gDNA mutations
## Global overview
### Overlaps between conditions
```{r}
knitr::include_graphics("../pdf/gDNA_conditions_overlaps.pdf")
```
### Mutation counts per condition and between conditions
```{r}
knitr::include_graphics("../pdf/gDNA_mutations_counts.pdf")
```
## Allele frequency (specific vs enriched)
```{r}
knitr::include_graphics("../pdf/gDNA_allele_freq_scatterplots.pdf")
```
## Overlaps between mutations shared between several conditions
```{r}
knitr::include_graphics("../pdf/gDNA_overlaps_mutation_shared_several_conditions.pdf")
```
## Mutations features
### All mutations features
```{r}
knitr::include_graphics("../pdf/gDNA_mut_features.pdf")
```
### Exonic mutations features
```{r}
knitr::include_graphics("../pdf/gDNA_exonic_mut_features.pdf")
```
## Gene centric analysis
### Overlapping mutant genes across conditions
```{r}
knitr::include_graphics("../pdf/gDNA_mutant_genes_overlap.pdf")
```
### Go of mutated genes (exonic mutations)
```{r}
knitr::include_graphics("../pdf/gDNA_mutant_genes_GO.pdf")
```
### Mutations impact on FC
```{r}
knitr::include_graphics("../pdf/gDNA_mutations_impact_FC.pdf")
```
## Mutations hotspots
```{r}
knitr::include_graphics("../pdf/gDNA_hostspots.pdf")
```




