---
title: "Epignetic Cancer project"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, out.width = "800px", out.height = "500px")
```

## 1/ RNA-Seq (only focused on GFP- system for now!)
Alignment statistics

```{r}
knitr::include_graphics("../pdf/Figures/Alignment_statistics_RNA.pdf")
```

GFP reads (genetic system validation)

```{r}
knitr::include_graphics("../pdf/Figures/GFP_norm_counts.pdf")
```

PCC replicates

```{r}
knitr::include_graphics("../pdf/Figures/PCC_replicates_RNA.pdf")
```

PCA replicates (QC, norm counts)

```{r}
knitr::include_graphics("../pdf/Figures/PCA_counts_RNA.pdf")
```

PCA conditions (compare conditions, log2FC)

```{r}
knitr::include_graphics("../pdf/Figures/PCA_log2FC_RNA.pdf")
```

MA plots (diff genes)

```{r}
knitr::include_graphics("../pdf/Figures/MA_plots_RNA.pdf")
```

Alluvial plots (transitions between conditions)

```{r}
knitr::include_graphics("../pdf/Figures/alluvial_plot_timecourse_RNA.pdf")
```

Gene overlaps between conditions

```{r}
knitr::include_graphics("../pdf/Figures/upsetplot_overlapping_genes_RNA.pdf")
```

Heatmap logFC genes of interest

```{r}
knitr::include_graphics("../pdf/Figures/Heatmap_FC_GOF.pdf")
```

GO up and down genes per condition

```{r}
knitr::include_graphics("../pdf/Figures/GO_up_down_genes_GFP-.pdf")
```

## 2/ CUTNRUN

PCC between replicates (QC)

```{r}
knitr::include_graphics("../pdf/cutnrun/PCC_cutnrun_replicates.pdf")
```

MA plots (diff regions)

```{r}
knitr::include_graphics("../pdf/cutnrun/MA_plots.pdf")
```

## 3/ Clustering based on RNA-Seq

Clustering figure
Col1 -> Clustering heatmaps
Col2 -> chromhmm SA2020
        PH peaks type SA2020
        FPKM W18 condition
Col3 -> STRING network of interacting proteins, colored by cluster
Col4 -> GO/cluster, depending on PRC1 binding (SA2020 assignment)
Col5 -> Motif enrichment/cluster, depending on PRC1 binding (SA2020 assignment)

```{r}
knitr::include_graphics("../pdf/Figures/Clustering_GFP-_system_RNA.pdf")
```

Plot cluster features depending on PRC1 binding (SA2020 assignment)

```{r}
knitr::include_graphics("../pdf/Figures/Cluster_PRC1_bound_unbound_RNA.pdf")
```

MA plots (diff regions)

```{r}
knitr::include_graphics("../pdf/Figures/Cluster_PRC1_bound_unbound_RNA.pdf")
```

4/ RECOVERY vs NO RECOVERY</b>
Defined as K27me+ & PRC1 bound (SA2020 assignment) &:
PHD9  -> log2FoldChange_PHD9_vs_PH29<(-1) & padj_PHD9_vs_PH29<(0.05) --> RECOVERY
         else if log2FoldChange_PHD9_vs_WKD>0 -------------------------> NORECOVERY
PHD11 -> log2FoldChange_PHD11_vs_PH29<(-1) & padj_PHD11_vs_PH29<(0.05) -> RECOVERY
         else if log2FoldChange_PHD11_vs_WKD>0 -------------------------> NORECOVERY
Table

FPKM, overlaps between then and with clusters

```{r}
knitr::include_graphics("../pdf/Figures/compare_RECOVERY_vs_transcriptome_cl2&5.pdf")
```

Screenshots CUTNRUN & RNA
blue: PH18, green: PHD11, magenta: PH29

```{r}
knitr::include_graphics("../pdf/Figures/RECOVERY_vs_NOT_screenshot.pdf")
```

Average tracks PH18/ctl conditions centered on PROMOTERS

```{r}
knitr::include_graphics("../pdf/Figures/PH18_CUTNRUN_enrich_revert_vs_not.pdf")
```

Average tracks PH18/ctl conditions centered on most proximal ATAC peaks (REs)

```{r}
knitr::include_graphics("../pdf/Figures/PH18_CUTNRUN_enrich_REs_revert_vs_not.pdf")
```

Average tracks between conditions centered on promoters

```{r}
knitr::include_graphics("../pdf/Figures/compare_CUTNRUN_cditions_revert_vs_not.pdf")
```

Motif enrichment at proximal ATAC peaks (REs) and PROMOTERS

```{r}
knitr::include_graphics("../pdf/Figures/motif_enrich_recovery_vs_no.pdf")
```

## 5/ cl2 (NO RECOVERY) vs cl5 (RECOVERY)
In my first attempt in using these clusters, we did not see much difference
But I did not filter the genes for being PRC1 bound and K27me3+!
Now it seems promising

Average tracks PH18/ctl conditions centered on PROMOTERS

```{r}
knitr::include_graphics("../pdf/Figures/PH18_CUTNRUN_enrich_cl2_vs_cl5.pdf")
```

Average tracks PH18/ctl conditions centered on most proximal PH peaks (REs)

```{r}
knitr::include_graphics("../pdf/Figures/PH18_CUTNRUN_enrich_REs_cl2_vs_cl5.pdf")
```

Average tracks between conditions centered on promoters

```{r}
knitr::include_graphics("../pdf/Figures/compare_CUTNRUN_cditions_cl2_vs_cl5.pdf")
```

Average tracks between conditions centered on promoters

```{r}
knitr::include_graphics("../pdf/Figures/compare_CUTNRUN_cditions_cl2_vs_cl5.pdf")
```

Motif enrichment at PROMOTERS, proximal ATAC peaks (REs) and proximal PH peaks (ED ChIP-Seq)

```{r}
knitr::include_graphics("../pdf/Figures/motif_enrich_cl2_vs_cl5.pdf")
```
